File: Dynamic Programming/Longest Common Subsequence.java
class Solution {
    static int[][] tb;
    public int longestCommonSubsequence(String s1, String s2) {
        int x = s1.length();
        int y = s2.length();
        tb = new int[x+1][y+1];
        for (int[] row: tb) {
            Arrays.fill(row, 0);
        }
        for(int i = 1; i<x+1; i++) {
            for (int j = 1; j < y + 1; j++) {
                if(s1.charAt(i-1) == s2.charAt(j-1)) {
                    tb[i][j] = 1 + tb[i-1][j-1];
                } else {
                    tb[i][j] = Math.max(tb[i][j-1], tb[i-1][j]);
                }
            }
        }
        return tb[x][y];
    }
}

File: Dynamic Programming/LCS/Longest Common Subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string s1, string s2) {
        int x = s1.length();
        int y = s2.length();

        int dp[x+1][y+1];
        memset(dp, 0, sizeof(dp));

        for (int i = 1; i < x+1; i++) {
            for (int j = 1; j < y+1; j++) {
                if (s1[i-1] == s2[j-1]) {
                    dp[i][j] = 1 + dp[i-1][j-1];
                } else {
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
                }
            }
        }
        return dp[x][y];
    }
};
